ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/API/Types/InvalidBlockPunishment.hs

```diff
@@ -63,27 +62,23 @@ data PeerSentAnInvalidBlockException = PeerSentAnInvalidBlockException
 
 instance Exn.Exception PeerSentAnInvalidBlockException
 
--- | Allocate a stateful punishment that performs the given punishment unless
--- the given header is better than the previous invocation
-mkUnlessImproved :: forall proxy m blk.
+-- | Allocate a stateful punishment that performs the given punishment if the
+-- given header does not satisfy the diffusion pipelining criterion.
+mkForDiffusionPipelining :: forall m blk.
      ( IOLike m
-     , NoThunks (SelectView (BlockProtocol blk))
-     , Ord      (SelectView (BlockProtocol blk))
+     , BlockSupportsDiffusionPipelining blk
      )
-  => proxy blk
-  -> STM m (   SelectView (BlockProtocol blk)
+  => STM m (   BlockConfig blk
+            -> Header blk
             -> InvalidBlockPunishment m
             -> InvalidBlockPunishment m
            )
-mkUnlessImproved _prx = do
-    var <- newTVar (NoLastInvalidTentative :: TentativeState blk)
-    pure $ \new punish -> InvalidBlockPunishment $ \invalidity -> join $ atomically $ do
-      io <- readTVar var <&> \case
-        NoLastInvalidTentative   -> pure ()
-        LastInvalidTentative old -> unless (new > old) $ do
-          enact punish invalidity
-      writeTVar var $ LastInvalidTentative new
-      pure io
+mkForDiffusionPipelining = do
+    var <- newTVar (initialTentativeHeaderState (Proxy @blk))
+    pure $ \cfg new punish -> InvalidBlockPunishment $ \invalidity -> join $ atomically $ do
+      readTVar var <&> updateTentativeHeaderState cfg new >>= \case
```

Comment:
I don't think this line is easy to read, since the operator fixities are not extremely common.

I'd suggest binding a `mbSt'` var, for example, to make it more legible.